### PR TITLE
Enable Mypy checking for scheduler.py

### PR DESCRIPTION
**Pull request by ipiszy**
ATT, add type annotations and type assertions to pass Mypy checks.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -52,8 +52,12 @@ def get_name(self):
 def fuse(node1: "BaseSchedulerNode", node2: "BaseSchedulerNode"):
     if node1.is_foreach():
         return ForeachKernelSchedulerNode.fuse(node1, node2)
-    else:
+    elif isinstance(node1, SchedulerNode) and isinstance(node2, SchedulerNode):
```
Comment (eellison) [inline torch/_inductor/scheduler.py:55]: I think this might be semantics changing ?

**Review by eellison — COMMENTED** [review]
Still have test failures

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -2132,6 +2132,9 @@ def get_reads(self):
     def realize(self):
         pass
 
+    def should_allocate(self):
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/ir.py:2572]: Yeah `InputBuffer` doesn't have `should_allocate` defined.

**Comment by ipiszy** [comment]
Sorry for the delay, reviving this PR since https://github.com/pytorch/pytorch/pull/108138 is not active.

**Comment by int3** [comment]
Looks like there are a bunch of test errors

File: torch/_inductor/dependencies.py [hunk 3]
```diff
@@ -131,7 +131,7 @@ class ReadWrites:
     index_exprs: Set[IndexExprDep]
     range_vars: Optional[List[sympy.Expr]] = None
     var_ranges: Optional[VarRanges] = None
-    op_counts: typing.Counter[Any] = None  # type: ignore[assignment]
+    op_counts: typing.Counter[Dict[str, int]] = None  # type: ignore[assignment]
```
Comment (int3) [inline torch/_inductor/dependencies.py:134]: I think the `type: ignore` can be removed if you make this an `Optional` type

Also, are you sure this isn't a `Counter[str]`? A Counter of Dict seems a bit odd

File: torch/_inductor/scheduler.py [hunk 4]
```diff
@@ -180,7 +182,7 @@ def used_or_aliased_buffer_names(self) -> Set[str]:
                 # needed to avoid deallocating aliased buffer
                 # if there are still uses of aliases ahead
                 if isinstance(layout, ir.AliasedLayout):
-                    used_names.add(layout.view.data.get_name())
+                    used_names.add(layout.view.get_name())
```
Comment (int3) [inline torch/_inductor/scheduler.py:185]: hm why this change

File: torch/_inductor/scheduler.py [hunk 5]
```diff
@@ -207,7 +209,7 @@ def prune_redundant_deps(self, name_to_fused_node):
         In essence this enforces an ordering on fusions. As fusions occur, prunable stardeps will
         be incrementally removed, enabling other fusions, ensuring they are fused in order.
         """
-        name_to_dep_count = collections.Counter()
+        name_to_dep_count: Counter[Dict[str, int]] = collections.Counter()
```
Comment (int3) [inline torch/_inductor/scheduler.py:212]: like above, I think this should probably be `Counter[str]`

File: torch/_inductor/scheduler.py [hunk 6]
```diff
@@ -296,10 +298,11 @@ def decide_inplace_update(self):
             ordered_reads = sorted(self.read_writes.reads, key=lambda x: x.name)
 
             for read in ordered_reads:
-                input_node: BaseSchedulerNode = self.scheduler.name_to_node.get(
-                    read.name
-                )
+                input_node: Optional[
+                    BaseSchedulerNode
+                ] = self.scheduler.name_to_node.get(read.name)
```
Comment (int3) [inline torch/_inductor/scheduler.py:303]: can we type `name_to_node` instead of typing this local variable? IMO we should avoid local variable types where possible because mypy can infer them

File: torch/_inductor/scheduler.py [hunk 7]
```diff
@@ -728,7 +744,8 @@ class FusedSchedulerNode(BaseSchedulerNode):
     @classmethod
     def fuse(cls, node1: BaseSchedulerNode, node2: BaseSchedulerNode):
         assert node1.scheduler is node2.scheduler
-        return cls(node1.scheduler, node1.get_nodes() + node2.get_nodes())
+        assert isinstance(node1, SchedulerNode) and isinstance(node2, SchedulerNode)
+        return cls(node1.scheduler, node1.get_nodes() + node2.get_nodes())  # type: ignore[arg-type]
```
Comment (int3) [inline torch/_inductor/scheduler.py:748]: just wondering, why is `type: ignore[arg-type]` needed here?

File: torch/_inductor/scheduler.py [hunk 8]
```diff
@@ -1206,7 +1226,7 @@ def compute_dependencies(self):
         Create dependency edges between nodes, handling aliasing and
         mutation properly.
         """
-        name_to_users = collections.defaultdict(list)
+        name_to_users: Dict[str, List] = collections.defaultdict(list)
```
Comment (int3) [inline torch/_inductor/scheduler.py:1229]: should be `typing.DefaultDict`

File: torch/_inductor/scheduler.py [hunk 9]
```diff
@@ -969,10 +988,8 @@ def __init__(
         else:
             self.scheduler = scheduler
             self.snodes = nodes
-            self.node = None
-
-            self.node = None
-            self.users = None
+            self.node = None  # type: ignore[assignment]
```
Comment (int3) [inline torch/_inductor/scheduler.py:991]: can we make BaseSchedulerNode.node an `Optional` type to avoid this `type: ignore`?

File: torch/_inductor/scheduler.py [hunk 10]
```diff
@@ -969,10 +988,8 @@ def __init__(
...
+            self.users = None  # type: ignore[assignment]
```
Comment (int3) [inline torch/_inductor/scheduler.py:992]: I think `self.users` is an Optional type, so this `type: ignore` shouldn't be necessary

File: torch/_inductor/scheduler.py [hunk 11]
```diff
@@ -792,7 +811,7 @@ def used_or_aliased_buffer_names(self) -> Set[str]:
         return set.union(*[x.used_or_aliased_buffer_names() for x in self.snodes])
 
     def get_nodes(self) -> List[BaseSchedulerNode]:
-        return self.snodes
+        return self.snodes  # type: ignore[return-value]
```
Comment (int3) [inline torch/_inductor/scheduler.py:814]: why is this necessary?

File: torch/_inductor/scheduler.py [hunk 12]
```diff
@@ -1524,7 +1547,7 @@ def found_path(node):
                 else:
                     # continue DFS of new ancestors introduced by the fusion
                     return bool(combined_names & node.ancestors) or any(
-                        found_path(self.name_to_fused_node[n])
+                        found_path(self.name_to_fused_node[n])  # type: ignore[index]
```
Comment (int3) [inline torch/_inductor/scheduler.py:1550]: why is this necessary?

File: torch/_inductor/scheduler.py [hunk 13]
```diff
@@ -88,7 +88,9 @@ def __init__(self, scheduler: "Scheduler", node: ir.Buffer):
         self.ancestors: Optional[Set[str]] = None
         self.min_order: Optional[int] = None
         self.max_order: Optional[int] = None
```
Comment (int3) [inline torch/_inductor/scheduler.py:90]: I guess we are declaring `min_order` and `max_order` here (even though the BaseSchedulerNode methods don't use them) because we want to be able to access them on instances of BaseSchedulerNode. But it doesn't seem like we ever actually expect them to have a `None` value. I *think* we could hack around this by declaring uninitialized fields, similar to what I did here: https://github.com/pytorch/pytorch/blob/fa7e5170f3c0b37552921fac6516aa513db3733d/torch/_inductor/ir.py#L309-L322

**Review by int3 — COMMENTED** [review]
I added some suggestions to try and minimize the number of `type: ignore` comments and `assert` statements

File: torch/_inductor/scheduler.py (hunk 13 above)
Comment (ipiszy) [reply torch/_inductor/scheduler.py]: yeah makes sense.

File: torch/_inductor/scheduler.py (hunk 6 above)
Comment (ipiszy) [reply torch/_inductor/scheduler.py:303]: Added type hint to `name_to_node`. I feel we still want to keep type hint of input node to improve readability.

File: torch/_inductor/scheduler.py (hunk 7 above)
Comment (ipiszy) [reply torch/_inductor/scheduler.py:749]: `FusedSchedulerNode` expects `List[SchedulerNode]` while `get_nodes()` returns `List[BaseSchedulerNode]`.

File: torch/_inductor/scheduler.py (hunk 11 above)
Comment (ipiszy) [reply torch/_inductor/scheduler.py:815]: `self.snodes` is `List[SchedulerNode]`

File: torch/_inductor/scheduler.py (hunk 10 above)
Comment (ipiszy) [reply torch/_inductor/scheduler.py]: `super().__init__()` is not called so it needs to be re-declared. Anyway added type hints.

File: torch/_inductor/scheduler.py (hunk 12 above)
Comment (ipiszy) [reply torch/_inductor/scheduler.py:1551]: mypy ignores the assertion added at line 1530. It thinks `self.name_to_fused_node` can be None.

File: torch/_inductor/scheduler.py (hunk 4 above)
Comment (ipiszy) [reply torch/_inductor/scheduler.py:185]: Otherwise, there will be an error "IRNode has no attribute data".

**Review by ipiszy — COMMENTED** [review]
Thanks @int3 !

File: torch/_inductor/scheduler.py (hunk 7 above)
Comment (int3) [reply torch/_inductor/scheduler.py:752]: I see. Looks like `FusedSchedulerNode` and `ForeachKernelSchedulerNode` share the same constructor. Given that it seems to be raising type errors in multiple places (see my comment below re `ForeachKernelSchedulerNode`), maybe it's not being correctly typed to begin with...

**Review by int3 — COMMENTED** [review]
More assert-removal suggestions... I think I got them all this time 😅

File: torch/_inductor/scheduler.py [hunk 14]
```diff
@@ -638,6 +653,7 @@ def get_ranges(self):
         return self._sizes
 
     def is_reduction(self):
+        assert isinstance(self.node, ir.ComputedBuffer)
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/scheduler.py:677]: I actually feel that it's better to keep this assert. IMO this piece of code is a bit hard to follow as node type is unclear. Also if self.node is not ir.ComputedBuffer the next statement will also fail, and adding an assertion message here can give clear error messages.

File: torch/_inductor/scheduler.py [hunk 15]
```diff
@@ -1846,6 +1871,7 @@ def codegen_extern_call(self, scheduler_node: ExternKernelSchedulerNode):
             scheduler_node.decide_inplace_update()
             scheduler_node.allocate()
         node = scheduler_node.node
+        assert isinstance(node, ir.ExternKernel)
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/scheduler.py:2043]: Similarly, I feel it's better to keep this assert to improve readability and give clear error messages.

File: torch/_inductor/scheduler.py (hunk 4 above)
Comment (ipiszy) [reply torch/_inductor/scheduler.py]: Updated view type hints in ir.py instead..

**Review by ipiszy — COMMENTED** [review]
Thanks @int3 !

File: torch/_inductor/scheduler.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (int3) [reply torch/_inductor/scheduler.py]: This boils down to whether we are thinking of readability in code review or readability in the IDE, I guess :) Personally I think statically encoding the type information makes for clearer code because it's more clear why we expect `self.node` to be a certain type. But I guess this approach is clean enough...

**Review by int3 — APPROVED** [review]
Thanks @ipiszy!

File: torch/_inductor/scheduler.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/scheduler.py]: I see your point. Tried it, it caused more errors because `ForEachSchedulerNode` and `FusedSchedulerNode` don't have node defined.  I feel it brings more troubles than gains..

[the remark this answers is not in the document]
Comment (int3) [reply torch/_inductor/scheduler.py]: Ah fair enough. Thanks for trying it!
